File: extract_claims_to_csv.py
#!/usr/bin/env python3
"""
Medical Claims (EOB) PDF → CSV extractor (prints-friendly DMBA-style).

What it does
------------
- Opens a "print-friendly" Explanation of Benefits (EOB) / claims history PDF.
- Extracts per-page header fields (Claim, Patient, Health Plan, Participant, Participant Id,
  Date Entered, Date Paid, Provider) from lines where labels often appear *without colons*
  (e.g., "Patient <name> Health Plan <plan>").
- Extracts each service line that appears as a single line:
      02/05/2025 OFFICE VISIT $223.00 $0.00 $102.36 B6 N3
- Writes one CSV containing every service line from all pages.

Dependencies
------------
- Python 3.9+
- pdfplumber  (install:  pip install pdfplumber)

Usage
-----
    python extract_claims_to_csv.py PrintFriendlyEOB.pdf history.csv
"""

import argparse
import csv
import re
from decimal import Decimal, InvalidOperation
from typing import Dict, List, Optional

import pdfplumber

# ---------- Currency + service-line parsing ----------
CURRENCY = r"(?:\$?(?:\d{1,3}(?:,\d{3})*|\d+)(?:\.\d{2})?)"

SERVICE_ROW_RE = re.compile(
    rf"""^
    (?P<ServiceDate>\d{{2}}/\d{{2}}/\d{{4}})\s+     # 02/05/2025
    (?P<Service>.*?)\s+                             # OFFICE VISIT (lazy)
    (?P<ProviderBilled>{CURRENCY})\s+               # $223.00
    (?P<DMBA_Paid>{CURRENCY})\s+                    # $0.00
    (?P<YourResp>{CURRENCY})\s+                     # $102.36
    (?P<MessageCodes>[A-Z0-9 ]{{1,40}})             # B6 N3
    $""",
    re.VERBOSE
)

def _money_to_str(val: str) -> str:
    """Normalize currency to plain 2-decimal text (remove $ and commas)."""
    s = val.replace(",", "").strip()
    if s.startswith("$"):
        s = s[1:]
    try:
        return f"{Decimal(s):.2f}"
    except (InvalidOperation, ValueError):
        return s

# ---------- Header extraction tuned to the sample you showed ----------
# We allow either coloned or non-coloned label variants.
HEADER_PATTERNS = [
    ("Claim", re.compile(r"\bClaim\s*:?\s*(T\d{7,})")),
    ("Patient", re.compile(r"\bPatient\s+(.+?)\s+Health\s*Plan\b", re.S)),
    ("Health Plan", re.compile(r"\bHealth\s*Plan\s+(.+?)\s+(?:Participant|Date\s*Entered)\b", re.S)),
    ("Participant", re.compile(r"\bParticipant\s+(.+?)\s+Date\s*Entered\b", re.S)),
    ("Participant Id", re.compile(r"\bParticipant\s*Id\s+([0-9]+)\b")),
    ("Date Entered", re.compile(r"\bDate\s*Entered\s+(\d{2}/\d{2}/\d{4})\b")),
    ("Date Paid", re.compile(r"\bDate\s*Paid\s+(\d{2}/\d{2}/\d{4})\b")),
    ("Provider", re.compile(r"\bProvider\s+(.+?)(?:\n|$)")),
]

def extract_header_fields(page_text: str) -> Dict[str, Optional[str]]:
    header = {k: None for k, _ in HEADER_PATTERNS}
    for key, pat in HEADER_PATTERNS:
        m = pat.search(page_text)
        if m:
            header[key] = " ".join(m.group(1).split())
    return header

def extract_service_lines(page_text: str) -> List[Dict[str, str]]:
    rows: List[Dict[str, str]] = []
    for raw in page_text.splitlines():
        line = raw.strip()
        if not line:
            continue
        m = SERVICE_ROW_RE.match(line)
        if m:
            rows.append({
                "Service Date": m.group("ServiceDate"),
                "Services Provided": m.group("Service").strip(),
                "Provider Billed ($)": _money_to_str(m.group("ProviderBilled")),
                "DMBA Paid ($)": _money_to_str(m.group("DMBA_Paid")),
                "Your Responsibility ($)": _money_to_str(m.group("YourResp")),
                "Message Codes": m.group("MessageCodes").strip(),
            })
    return rows

def parse_pdf_to_rows(pdf_path: str) -> List[Dict[str, str]]:
    all_rows: List[Dict[str, str]] = []
    with pdfplumber.open(pdf_path) as pdf:
        for page in pdf.pages:
            # Keep newlines (row boundaries), normalize internal spacing.
            text = page.extract_text() or ""
            text = text.replace("\xa0", " ")
            cleaned_lines = [re.sub(r"[ \t]{2,}", " ", ln).strip() for ln in text.splitlines()]
            page_text = "\n".join(cleaned_lines)

            header = extract_header_fields(page_text)
            service_rows = extract_service_lines(page_text)

            for r in service_rows:
                stamped = {
                    "Claim": header.get("Claim") or "",
                    "Patient": header.get("Patient") or "",
                    "Health Plan": header.get("Health Plan") or "",
                    "Participant": header.get("Participant") or "",
                    "Participant Id": header.get("Participant Id") or "",
                    "Date Entered": header.get("Date Entered") or "",
                    "Date Paid": header.get("Date Paid") or "",
                    "Provider": header.get("Provider") or "",
                }
                stamped.update(r)
                all_rows.append(stamped)
    return all_rows

def write_csv(rows: List[Dict[str, str]], out_path: str) -> None:
    fieldnames = [
        "Claim", "Patient", "Health Plan", "Participant", "Participant Id",
        "Date Entered", "Date Paid", "Provider",
        "Service Date", "Services Provided",
        "Provider Billed ($)", "DMBA Paid ($)", "Your Responsibility ($)",
        "Message Codes",
    ]
    with open(out_path, "w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=fieldnames)
        w.writeheader()
        for r in rows:
            w.writerow({k: r.get(k, "") for k in fieldnames})

def main():
    ap = argparse.ArgumentParser(description="Extract medical claims service lines from a print-friendly EOB PDF into CSV.")
    ap.add_argument("pdf", help="Path to the print-friendly EOB PDF")
    ap.add_argument("csv", help="Path to output CSV file")
    args = ap.parse_args()

    rows = parse_pdf_to_rows(args.pdf)
    write_csv(rows, args.csv)
    print(f"Extracted {len(rows)} service line(s) to: {args.csv}")

if __name__ == "__main__":
    main()

File: extract_claims_to_csv.mjs
#!/usr/bin/env node
/**
 * Medical Claims (EOB) PDF → CSV extractor (DMBA-style) using pdfjs-dist.
 *
 * This version disables font loading (disableFontFace: true) to avoid font warnings.
 * That’s safe because we only need text content, not font rendering.
 *
 * Usage:
 *   node extract_claims_to_csv.mjs PrintFriendlyEOB.pdf history.csv
 *
 * Deps:
 *   npm i pdfjs-dist
 */

import fs from "fs";
import path from "path";
import * as pdfjsLib from "pdfjs-dist/legacy/build/pdf.mjs";

// ------------- CLI -------------
if (process.argv.length < 4) {
  console.error("Usage: node extract_claims_to_csv.mjs <input.pdf> <output.csv>");
  process.exit(1);
}
const inputPdf = process.argv[2];
const outputCsv = process.argv[3];

// Disable web worker in Node and quiet logs
pdfjsLib.GlobalWorkerOptions.disableWorker = true;
if (pdfjsLib.setVerbosityLevel && pdfjsLib.VerbosityLevel) {
  pdfjsLib.setVerbosityLevel(pdfjsLib.VerbosityLevel.ERROR);
}

// ------------- Helpers -------------
function normalizeSpacesKeepNewlines(s) {
  return s
    .replace(/\u00a0/g, " ")
    .split("\n")
    .map((ln) => ln.replace(/[ \t]{2,}/g, " ").trimEnd())
    .join("\n");
}

function toCsvLine(values) {
  return values
    .map((v) => {
      const s = String(v ?? "");
      return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
    })
    .join(",");
}

function writeCsv(rows, outPath) {
  const headers = [
    "Claim", "Patient", "Health Plan", "Participant", "Participant Id",
    "Date Entered", "Date Paid", "Provider",
    "Service Date", "Services Provided",
    "Provider Billed ($)", "DMBA Paid ($)", "Your Responsibility ($)",
    "Message Codes",
  ];
  const out = [toCsvLine(headers)]
    .concat(rows.map((r) => toCsvLine(headers.map((h) => r[h] ?? ""))))
    .join("\n") + "\n";
  fs.writeFileSync(outPath, out, "utf8");
}

function moneyToStr(val) {
  if (!val) return "";
  let s = val.replace(/,/g, "").trim();
  if (s.startsWith("$")) s = s.slice(1);
  const num = Number.parseFloat(s);
  return Number.isFinite(num) ? num.toFixed(2) : s;
}

// ---- Patterns tuned to this EOB layout ----
const CURRENCY = String.raw`(?:\$?(?:\d{1,3}(?:,\d{3})*|\d+)(?:\.\d{2})?)`;

const SERVICE_ROW_RE = new RegExp(
  String.raw`^` +
    String.raw`(?<ServiceDate>\d{2}\/\d{2}\/\d{4})\s+` +   // date
    String.raw`(?<Service>.*?)\s+` +                       // desc (lazy)
    String.raw`(?<ProviderBilled>${CURRENCY})\s+` +
    String.raw`(?<DMBA_Paid>${CURRENCY})\s+` +
    String.raw`(?<YourResp>${CURRENCY})\s+` +
    String.raw`(?<MessageCodes>[A-Z0-9 ]{1,40})` +
  String.raw`$`
);

const HEADER_PATTERNS = [
  ["Claim",          /\bClaim\s*:?\s*(T\d{7,})/s],
  ["Patient",        /\bPatient\s+(.+?)\s+Health\s*Plan\b/s],
  ["Health Plan",    /\bHealth\s*Plan\s+(.+?)\s+(?:Participant|Date\s*Entered)\b/s],
  ["Participant",    /\bParticipant\s+(.+?)\s+Date\s*Entered\b/s],
  ["Participant Id", /\bParticipant\s*Id\s+([0-9]+)\b/s],
  ["Date Entered",   /\bDate\s*Entered\s+(\d{2}\/\d{2}\/\d{4})\b/s],
  ["Date Paid",      /\bDate\s*Paid\s+(\d{2}\/\d{2}\/\d{4})\b/s],
  ["Provider",       /\bProvider\s+(.+?)(?:\n|$)/s],
];

function extractHeaderFields(pageText) {
  const header = {};
  for (const [key, rx] of HEADER_PATTERNS) {
    const m = pageText.match(rx);
    header[key] = m && m[1] ? m[1].replace(/\s+/g, " ").trim() : "";
  }
  return header;
}

function extractServiceLines(pageText) {
  const rows = [];
  for (const raw of pageText.split("\n")) {
    const line = raw.trim();
    if (!line) continue;
    const m = line.match(SERVICE_ROW_RE);
    if (m?.groups) {
      rows.push({
        "Service Date": m.groups.ServiceDate,
        "Services Provided": m.groups.Service.trim(),
        "Provider Billed ($)": moneyToStr(m.groups.ProviderBilled),
        "DMBA Paid ($)": moneyToStr(m.groups.DMBA_Paid),
        "Your Responsibility ($)": moneyToStr(m.groups.YourResp),
        "Message Codes": m.groups.MessageCodes.trim(),
      });
    }
  }
  return rows;
}

/** Rebuild lines from pdf.js text items by grouping items with similar Y and sorting by X. */
function pageTextFromItems(textContent, { yRound = 1, sortY = "desc" } = {}) {
  const linesByY = new Map(); // y -> [{x,str},...]
  for (const it of textContent.items) {
    const [,, , , x, y] = it.transform; // a,b,c,d,e(x),f(y)
    const yKey = yRound > 0 ? Math.round(y / yRound) * yRound : Math.round(y);
    const arr = linesByY.get(yKey) || [];
    arr.push({ x, str: it.str });
    linesByY.set(yKey, arr);
  }
  const ys = Array.from(linesByY.keys()).sort((a, b) => (sortY === "asc" ? a - b : b - a));
  const lines = ys.map((y) =>
    linesByY.get(y).sort((p, q) => p.x - q.x).map((p) => p.str).join(" ").replace(/\s{2,}/g, " ").trim()
  );
  return lines.join("\n");
}

// ------------- Main -------------
(async function main() {
  try {
    const data = new Uint8Array(fs.readFileSync(path.resolve(inputPdf)));

    const loadingTask = pdfjsLib.getDocument({
      data,
      isEvalSupported: false,
      disableFontFace: true,     // <-- key: do not load fonts (no warnings)
      cMapPacked: true,          // fine to keep; we don't set cMapUrl -> built-ins
    });
    const doc = await loadingTask.promise;

    const allRows = [];

    for (let i = 1; i <= doc.numPages; i++) {
      const page = await doc.getPage(i);
      const content = await page.getTextContent({ normalizeWhitespace: true });
      const pageRaw = pageTextFromItems(content, { yRound: 1, sortY: "desc" });
      const pageText = normalizeSpacesKeepNewlines(pageRaw);

      const header = extractHeaderFields(pageText);
      const serviceRows = extractServiceLines(pageText);

      for (const r of serviceRows) {
        allRows.push({
          "Claim": header["Claim"] || "",
          "Patient": header["Patient"] || "",
          "Health Plan": header["Health Plan"] || "",
          "Participant": header["Participant"] || "",
          "Participant Id": header["Participant Id"] || "",
          "Date Entered": header["Date Entered"] || "",
          "Date Paid": header["Date Paid"] || "",
          "Provider": header["Provider"] || "",
          ...r,
        });
      }
    }

    writeCsv(allRows, path.resolve(outputCsv));
    console.log(`Extracted ${allRows.length} service line(s) to: ${outputCsv}`);
  } catch (err) {
    console.error("Error:", err);
    process.exit(1);
  }
})();

